src/com/teamcautionrobotics/robot2018/Lift.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    public void stop() {
+        this.move(0);
+    }
+
+    public void setLevel(LiftLevel destinationLiftLevel) {
+        setHeight(destinationLiftLevel.height);
+    }
+
+    public void setHeight(double height) {
+        pidController.setSetpoint(height);
+        enablePID();
+    }
+
+    public LiftLevel getCurrentLiftLevel() {
+        return convertHeightToLiftLevel(getCurrentHeight());
+    }
+
+    public double getCurrentHeight() {
+        return liftEncoder.getDistance();
+    }
+
+    public double getDestinationHeight() {
+        return pidController.getSetpoint();
+    }
+
+    public LiftLevel getDestinationLiftLevel() {
+        return convertHeightToLiftLevel(getDestinationHeight());
+    }
+
+    public static LiftLevel convertHeightToLiftLevel(double height) {
+        LiftLevel convertedLiftLevel = LiftLevel.values()[0];
+        for (LiftLevel liftLevel : LiftLevel.values) {
+            double midpoint = (liftLevel.height + liftLevel.next().height) / 2;
+            convertedLiftLevel = liftLevel;
+            // OHHHHH WE'RE HALFWAY THERE! WHOAWHOA then move on to the next LiftLevel and try
+            // again; if not, break out of the loop and return the current liftPosition the for loop
+            // is on
+            if (height < midpoint) {
+                break;
+            }
+        }
+        return convertedLiftLevel;
+    }
+
+    public void enablePID() {
+        if (!pidController.isEnabled()) {
+            pidController.enable();
+        }
+    }
+
+    public void disablePID() {
+        if (pidController.isEnabled()) {
+            pidController.disable();
+        }
+    }
+
+    public void resetEncoder() {
+        liftEncoder.reset();
+    }
+
+    class LiftPIDSource implements PIDSource {
+
+        private PIDSourceType type;
+
+        public LiftPIDSource(PIDSourceType sourceType) {
+            type = sourceType;
+        }
+
+        @Override
+        public void setPIDSourceType(PIDSourceType sourceType) {
+            type = sourceType;
+        }
+
+        @Override
+        public PIDSourceType getPIDSourceType() {
+            return type;
+        }
+
+        @Override
+        public double pidGet() {
+            switch (type) {
+                case kDisplacement:
+                  return getCurrentHeight();
+                case kRate:
+                  return liftEncoder.getRate();
+                default:
+                  return 0.0;
+              }
+        }
+        
+    }
+
+    @Override
+    /**
+     * Not implemented. Always displacement pid source.
+     * 
+     * @see edu.wpi.first.wpilibj.PIDSource#setPIDSourceType(edu.wpi.first.wpilibj.PIDSourceType)
+     */
+    public void setPIDSourceType(PIDSourceType pidSource) {
+
+    }
+
+    @Override
+    public PIDSourceType getPIDSourceType() {
+        return PIDSourceType.kDisplacement;
+    }
+
+    @Override
+    public double pidGet() {
+        return getCurrentHeight();
+    }
```

Comment:
We should remove this implementation of PIDSource, since we implement it better in LiftPIDSource.